File: clients/android/Vertex/core/src/main/kotlin/ru/vertices/android/core/protocol/JoinMessage.kt
package ru.vertices.android.core.protocol

import kotlinx.serialization.SerialName
import kotlinx.serialization.Serializable

/**
 * Join handshake. Published by the client to `vpn/{exit}/control/join`.
 *
 * Field names match Go `cmd/exit` and Swift `VertexCore.JoinMessage` byte-for-byte
 * — `id_sig` is intentionally snake_case on the wire.
 */
@Serializable
data class JoinMessage(
    @SerialName("name")   val name: String,
    /** Base64 of the client's ephemeral X25519 public key (32 bytes raw). */
    @SerialName("dh")     val dh: String,
    /** Base64 of the client's persistent X25519 identity public key (32 bytes raw). */
    @SerialName("id")     val id: String? = null,
    /** Base64 of the HMAC-SHA256 identity proof. See [ru.vertices.android.core.crypto.IdentityKey.proof]. */
    @SerialName("id_sig") val idSig: String? = null,
)

File: clients/windows/src/Vertex.Core/Protocol/JoinMessage.cs
using System.Text.Json.Serialization;

namespace Vertex.Core.Protocol;

/// <summary>
/// Join handshake message published by the client to <c>vpn/{exit}/control/join</c>.
/// Field names match Go <c>cmd/exit</c>, Swift <c>JoinMessage</c>, and Kotlin
/// <c>JoinMessage</c> byte-for-byte — <c>id_sig</c> is intentionally snake_case
/// on the wire.
/// </summary>
public sealed record JoinMessage(
    [property: JsonPropertyName("name")]   string  Name,
    /// <summary>Base64 of the client's ephemeral X25519 public key (32 bytes raw).</summary>
    [property: JsonPropertyName("dh")]     string  Dh,
    /// <summary>Base64 of the client's persistent X25519 identity public key (32 bytes raw). Optional for backward compat.</summary>
    [property: JsonPropertyName("id"),     JsonIgnore(Condition = JsonIgnoreCondition.WhenWritingNull)] string? Id    = null,
    /// <summary>Base64 of the HMAC-SHA256 identity proof. See <c>IdentityKey.Proof</c>.</summary>
    [property: JsonPropertyName("id_sig"), JsonIgnore(Condition = JsonIgnoreCondition.WhenWritingNull)] string? IdSig = null);
